lib/matplotlib/axes/_base.py

```diff
@@ -1322,7 +1325,16 @@ def axis(self, *v, **kwargs):
                     self.autoscale_view(tight=True)
                     self.set_autoscale_on(False)
                     self.set_aspect('equal', adjustable='box', anchor='C')
-
+                elif s == 'square':
+                    self.set_aspect('equal', adjustable='box', anchor='C')
+                    self.set_autoscale_on(False)
+                    xlim = self.get_xlim()
+                    ylim = self.get_ylim()
+                    edge_size = max(np.diff(xlim), np.diff(ylim))
+                    self.set_xlim([xlim[0], xlim[0] + edge_size],
```

Comment:
It would be nice to set up a callback on the zoom signal to enforce this after a user zoom, but that may be a bit too involved (as you would need to keep track of the cid and remove it if a different aspect was set.